### DISCUSSION TITLE
Run multiple Xapp

### DISCUSSION THREAD

**Discussion by ying5690** [srsran/oran-sc-ric#15] [General]
Hi, I am trying to run multiple Xapp.  For example the kpm_mon_xapp.py and simple_rc_xapp.py.
I have added a new route to the routes.rtg file, so the file has 2 end points to 12050, as follows:

rte|12050|10.0.2.20:4560   # RIC_INDICATION
rte|12050|10.0.2.20:4561  # RIC_INDICATION, added new route

Xapp A is requesting via MY_RMR_PORT 4560, HTTP_SERVER_PORT 8091
Xapp B is requesting via MY_RMR_PORT 4561, HTTP_SERVER_PORT 8092

I am getting the error when trying to request for subscription from the RIC.

Could you please elaborate a bit more how to implement multiple Xapp? I did notice there is a note in the readme mentioning adding route and differentiate subscription id in the xapp. It would be great if more details could be provided. Thank you.

**Comment by pgawlowicz** [comment]
Hi @ying5690,
It should be possible to run multiple xapp with the `pr_multi_xapps` branch. 

Please try to run:
```
docker compose exec python_xapp_runner ./simple_mon_xapp.py
# or
# docker compose exec python_xapp_runner ./kpm_mon_xapp.py --kpm_report_style=5
```
and 
```
docker compose exec python_xapp_runner ./simple_rc_xapp.py
```
at the same time.

Note that if you try to run two xApps that subscribe for the KPM measurements the gnb will crash. The bug was already found and fixed and will be public with the next release. 

Best,
Piotr

**Reply by ying5690** [reply]
Thank you Piotr! Branch pr_multi_xapps works for me!

**Reply by OniReimu** [reply]
Hi @pgawlowicz , great to see this timely update though, I am not sure if I am the only that came across this issue: 

I tried running the example xApps one by one starting from `simple_mon_xapp.py` and then `kpm_mon_xapp.py`, and `simple_rc_xapp.py` and finally the `simple_xapp.py` needs an explicit assignment of its `http_server_port` and `rmr_port`, otherwise it would end up getting address already being used. The default `8090` and `4560` are not working.

**Reply by pgawlowicz** [reply]
Indeed, `simple_rc_xapp.py` and `simple_xapp.py` use the same ports, as I have added only additional routing entries for the `RIC_INDICATION` message, see here: https://github.com/srsran/oran-sc-ric/blob/main/ric/configs/routes.rtg

There is only one route entry for `RIC_CONTROL_ACK` and `RIC_CONTROL_FAILURE` with port 4560, and those messages are used by the mentioned xApps.

**Reply by OniReimu** [reply]
But then `simple_rc_xapp.py` would be unsubscribed when quitted, there shouldn't be any collision? Did I miss something?

**Reply by pgawlowicz** [reply]
yes, it should be freed after the xapp exits.

**Reply by OniReimu** [reply]
Mine was like quit `simple_rc_xapp.py` and rerun it, worked.
But quit `simple_rc_xapp.py` and run `simle_xapp.py` not working.

**Reply by pgawlowicz** [reply]
it is fixed on `fix_missing_config_param` branch
